pkg/vmprovider/providers/vsphere2/virtualmachine/backup.go

```diff
@@ -28,128 +40,149 @@ type VMDiskData struct {
 			ExtraConfig: ecToUpdate,
 		}); err != nil {
-			ctx.VMCtx.Logger.Error(err, "Failed to update VM ExtraConfig for backup")
+			opts.VMCtx.Logger.Error(err, "Failed to update VM ExtraConfig with latest backup data")
 			return err
 		}
 	}
 
 	return nil
 }
 
-func getDesiredVMKubeDataForBackup(
-	vm *vmopv1.VirtualMachine,
+func getDesiredKubeObjectsYAMLForBackup(
+	kubeObjects []client.Object,
 	ecMap map[string]string) (string, error) {
-	// If the ExtraConfig already contains the latest VM spec, determined by
-	// 'metadata.generation', return an empty string to skip the backup.
-	if ecKubeData, ok := ecMap[vmopv1.VMBackupKubeDataExtraConfigKey]; ok {
-		vmFromBackup, err := constructVMObj(ecKubeData)
-		if err != nil {
-			return "", err
-		}
-		if vmFromBackup.ObjectMeta.Generation >= vm.ObjectMeta.Generation {
-			return "", nil
+	// Check if the VM's ExtraConfig already contains the latest version of all
+	// objects, determined by each resource UID to resource version mapping.
+	var isLatestBackup bool
+	if ecKubeData, ok := ecMap[vmopv1.VMBackupKubeObjectsYAMLExtraConfigKey]; ok {
+		if resourceToVersion := tryGetResourceVersion(ecKubeData); resourceToVersion != nil {
+			isLatestBackup = true
+			for _, curObj := range kubeObjects {
+				if curObj.GetResourceVersion() != resourceToVersion[string(curObj.GetUID())] {
+					isLatestBackup = false
+					break
+				}
+			}
 		}
 	}
 
-	backupVM := vm.DeepCopy()
-	backupVM.Status = vmopv1.VirtualMachineStatus{}
-	backupVMYaml, err := yaml.Marshal(backupVM)
-	if err != nil {
-		return "", err
+	// All objects are up-to-date, return an empty string to skip the backup.
+	if isLatestBackup {
+		return "", nil
 	}
 
-	return util.EncodeGzipBase64(string(backupVMYaml))
+	// Backup the latest version of all objects with encoded and gzipped YAML.
+	// Use "---" as the separator between objects to allow for easy parsing.
+	marshaledStrs := []string{}
+	for _, obj := range kubeObjects {
+		marshaledYaml, err := yaml.Marshal(obj)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal object %s: %v", obj.GetName(), err)
+		}
+		marshaledStrs = append(marshaledStrs, string(marshaledYaml))
+	}
+
+	kubeObjectsYAML := strings.Join(marshaledStrs, "---\n")
+	return util.EncodeGzipBase64(kubeObjectsYAML)
 }
 
-func constructVMObj(ecKubeData string) (vmopv1.VirtualMachine, error) {
-	var vmObj vmopv1.VirtualMachine
-	decodedKubeData, err := util.TryToDecodeBase64Gzip([]byte(ecKubeData))
+func tryGetResourceVersion(ecKubeObjectsYAML string) map[string]string {
+	decoded, err := util.TryToDecodeBase64Gzip([]byte(ecKubeObjectsYAML))
 	if err != nil {
-		return vmObj, err
+		return nil
+	}
+
+	resourceVersions := map[string]string{}
+	kubeObjectsYAML := strings.Split(decoded, "---\n")
+	for _, objYAML := range kubeObjectsYAML {
+		if objYAML != "" {
+			objYAML = strings.TrimSpace(objYAML)
+			var obj map[string]interface{}
```

Comment:
Replace with `unstructured.Unstructured` which exposes methods to get UID and resource versions

example:
```golang
import (
    "k8s.io/apimachinery/pkg/runtime/serializer/yaml"
)


// decode YAML into unstructured.Unstructured
dec := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
_, gvk, err := dec.Decode([]byte(dsManifest), nil, obj)
```

Reply:
Nice! Updated as suggested.